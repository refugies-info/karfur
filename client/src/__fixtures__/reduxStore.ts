import { RootState } from "services/rootReducer";
import { SearchResultsState } from "services/SearchResults/searchResults.reducer";
import { UserFavoritesState } from "services/UserFavoritesInLocale/UserFavoritesInLocale.reducer";
import { activeThemesMock } from "./activeThemes";

const initialMockUserState = {
  user: null,
  admin: false,
  traducteur: false,
  expertTrad: false,
  contributeur: false,
  membreStruct: false,
  userId: "",
  userFetched: false,
  rolesInStructure: [],
};
const initialMockLangueState = {
  langues: [],
  languei18nCode: "fr",
  showLanguageModal: false,
  showLangModal: false,
};

const initialMockTranslationState = {
  translation: {
    initialText: {},
    translatedText: {},
  },
  translations: [],
};

const initialUserFavoritesState: UserFavoritesState = {
  favorites: [],
  showFavoriteModal: false
}

const initialMockSearchReults: SearchResultsState = {
  results: {
    dispositifs: [],
    demarches: [],
    dispositifsSecondaryTheme: []
  },
  query: {
    search: "",
    departments: [],
    themes: [],
    needs: [],
    age: [],
    frenchLevel: [],
    language: [],
    sort: "date",
    type: "all",
  }
}

export const initialMockStore = {
  user: initialMockUserState,
  langue: initialMockLangueState,
  activeDispositifs: [],
  tts: { ttsActive: false, showAudioSpinner: false },
  userStructure: null,
  selectedDispositif: null,
  loadingStatus: {},
  translation: initialMockTranslationState,
  activeStructures: [],
  selectedStructure: null,
  allDispositifs: [],
  allStructures: [],
  users: [],
  userFavorites: initialUserFavoritesState,
  userContributions: [],
  needs: [],
  themes: {
    activeThemes: activeThemesMock,
    inactiveThemes: []
  },
  searchResults: initialMockSearchReults
};
