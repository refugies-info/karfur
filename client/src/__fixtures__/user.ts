import { GetUserInfoResponse, UserStatus } from "api-types";
import moment from "moment";

const date = moment("2019-05-01 12:00:00");
export const testUser: GetUserInfoResponse = {
  username: "test",
  _id: "testObjectId",
  structures: [],
  contributions: [],
  phone: "",
  roles: [],
  email: "",
  selectedLanguages: [],
  status: UserStatus.ACTIVE,
  traductionsFaites: []
};

export const testUserWithRoles: GetUserInfoResponse = {
  ...testUser,
  structures: ["testObjectId"],
  roles: [
    { nom: "Admin", _id: "testObjectId", nomPublic: "Admin" },
    { nom: "Trad", _id: "testObjectId", nomPublic: "Trad" },
    {
      nom: "ExpertTrad",
      _id: "testObjectId",
      nomPublic: "ExpertTrad"
    },
    {
      nom: "Contrib",
      _id: "testObjectId",
      nomPublic: "Contrib"
    }
  ]
};
