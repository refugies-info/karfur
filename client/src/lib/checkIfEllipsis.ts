export const checkIfEllipsis = (element: HTMLElement) => element.offsetWidth < element.scrollWidth;
