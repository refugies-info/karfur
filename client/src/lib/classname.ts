export const cls = (...classNames: (string | boolean)[]) => classNames.filter(className => !!className).join(" ");
