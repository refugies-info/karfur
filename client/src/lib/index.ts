export const jsUcfirst = (string: string) =>
  string &&
  string.length > 1 &&
  string.charAt(0).toUpperCase() + string.slice(1, string.length);

export const jsLcfirst = (string: string) =>
  string &&
  string.length > 1 &&
  string.charAt(0).toLowerCase() + string.slice(1, string.length);

export const limitNbCaracters = (string: string, nbCaractersMax: number) =>
  string.substring(0, Math.min(string.length, nbCaractersMax)) +
  (string.length > nbCaractersMax ? "..." : "");

export const removeAccents = (str = "") => {
  var accents =
    "ÀÁÂÃÄÅàáâãäåÒÓÔÕÕÖØòóôõöøÈÉÊËèéêëðÇçÐÌÍÎÏìíîïÙÚÛÜùúûüÑñŠšŸÿýŽž";
  var accentsOut =
    "AAAAAAaaaaaaOOOOOOOooooooEEEEeeeeeCcDIIIIiiiiUUUUuuuuNnSsYyyZz";
  const chars = str.split("");
  let i, x;
  for (i = 0; i < chars.length; i++) {
    if ((x = accents.indexOf(chars[i])) !== -1) {
      chars[i] = accentsOut[x];
    }
  }
  return chars.join("");
};

// rules for password check :
// nbNumUpper = nbNumeric + nbUpper
// score 0 (cannot set password): length < 5
// score 0.5 : 7>length>=5 + nbNumUpper = 1 or 2 or0
// score 1.5 : 7>length >= 5 + nbNumUpper>=3 OR length>=7 + nbNumUpper<3
// score 2.5 : length >= 7 + nbNumUpper>=3
// score 3.5 : length >= 7 + nbNumUpper>=3 + at least 1 special caractere

// make sure to use same function in backend
export const computePasswordStrengthScore = (
  password: string
): { score: number } => {
  if (password.length < 5) {
    return { score: 0 };
  }

  const numerics = password.match(/\d+/g)?.join("");
  const nbNumeric = numerics ? numerics.length : 0;

  const upperCase = password.match(/[A-Z]/g)?.join("");
  const nbUpperCase = upperCase ? upperCase.length : 0;

  if (password.length < 7) {
    if (nbUpperCase + nbNumeric < 3) {
      return { score: 1 };
    }
    return { score: 2 };
  }

  if (nbNumeric + nbUpperCase < 3) {
    return { score: 2 };
  }
  const specialCar = password.match(/[!@#$&*]/g)?.join("");
  const nbSpecialCar = specialCar ? specialCar.length : 0;
  if (nbSpecialCar > 0) {
    return { score: 4 };
  }

  return { score: 3 };
};
