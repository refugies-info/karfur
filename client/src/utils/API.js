import axios from "axios";

import setAuthToken from "./setAuthToken";
import Swal from "sweetalert2";
import { logger } from "../logger";
import isInBrowser from "lib/isInBrowser";

const socket = null;
export { socket };


const burl = process.env.NEXT_PUBLIC_REACT_APP_SERVER_URL;

axios.withCredentials = true;

axios.interceptors.request.use(
  (request) => {
    return request;
  },
  (error) => {
    Swal.fire({
      type: "error",
      title: "Oops...",
      text: (error.response.data || {}).text || "",
      footer: "<i>" + error.message + "</i>",
      timer: 1500,
    });

    return Promise.reject(error);
  }
);

axios.interceptors.response.use(
  (response) => {
    // console.log(response)
    return response;
  },
  (error) => {
    if (error.response && error.response.status < 500) {
      if (error.response.data.data !== "no-alert") {
        Swal.fire({
          type: "error",
          title: "Oops...",
          text: (error.response.data || {}).text || "",
          footer: "<i>" + error.message + "</i>",
          timer: 1500,
        });
      }
    } else if (axios.isCancel(error)) {
      logger.error("Error: ", { error: error.message });
    }
    return Promise.reject(error);
  }
);

const CancelToken = axios.CancelToken;
let cancel;
// const source = CancelToken.source();
// const sourceToken = source.token

// export {source};

const getHeaders = () => {
  return {
    "Content-Type": "application/json",
    "x-access-token": localStorage.getItem("token") || undefined,
    "site-secret": process.env.NEXT_PUBLIC_REACT_APP_SITE_SECRET,
  };
}

export default {
  login: (user) => {
    const headers = getHeaders();
    return axios.post(burl + "/user/login", user, { headers });
  },
  signup: (send) => {
    const headers = getHeaders();
    return axios.post(burl + "/user/signup", send, { headers });
  },
  checkUserExists: (send) => {
    const headers = getHeaders();
    return axios.post(burl + "/user/checkUserExists", send, {
      headers,
    });
  },
  set_user_info: (user) => {
    const headers = getHeaders();
    return axios.post(burl + "/user/set_user_info", user, { headers });
  },
  get_users: (params = {}) => {
    const headers = getHeaders();
    return axios.post(burl + "/user/get_users", params, { headers });
  },
  get_user_info: (query) => {
    const headers = getHeaders();
    return axios.post(burl + "/user/get_user_info", query, {
      headers,
    });
  },
  updateUser: (query) => {
    const headers = getHeaders();
    return axios.post(burl + "/user/updateUser", query, {
      headers,
    })
  },
  changePassword: (query) => {
    const headers = getHeaders();
    return axios.post(burl + "/user/changePassword", query, {
      headers,
    })
  },
  reset_password: (query) => {
    const headers = getHeaders();
    return axios.post(burl + "/user/reset_password", query, {
      headers,
    });
  },
  set_new_password: (query) => {
    const headers = getHeaders();
    return axios.post(burl + "/user/set_new_password", query, {
      headers,
    });
  },
  addDispositif: (query) => {
    const headers = getHeaders();
    axios.post(burl + "/dispositifs/addDispositif", query, {
      headers,
    })
  },

  sendAdminImprovementsMail: (query) => {
    const headers = getHeaders();
    axios.post(burl + "/mail/sendAdminImprovementsMail", query, {
      headers,
    })
  },

  sendSubscriptionReminderMail: (query) => {
    const headers = getHeaders();
    axios.post(burl + "/mail/sendSubscriptionReminderMail", query, {
      headers,
    })
  },

  add_dispositif_infocards: (query) => {
    return axios.post(burl + "/dispositifs/add_dispositif_infocards", query, {
      headers,
    });
  },
  get_dispositif: (params = {}) => {
    return axios.post(burl + "/dispositifs/get_dispositif", params, {
      headers,
    });
  },
  count_dispositifs: (query) => {
    return axios.post(burl + "/dispositifs/count_dispositifs", query, {
      headers,
    });
  },
  updateDispositifReactions: (query) => {
    const headers = getHeaders();
    return axios.post(burl + "/dispositifs/updateDispositifReactions", query, {
      headers,
    })
  },

  updateDispositifStatus: (query) => {
    const headers = getHeaders();
    return axios.post(burl + "/dispositifs/updateDispositifStatus", query, {
      headers,
    })
  },

  updateDispositifTagsOrNeeds: (query) => {
    const headers = getHeaders();
    return axios.post(burl + "/dispositifs/updateDispositifTagsOrNeeds", query, {
      headers,
    })
  },

  modifyDispositifMainSponsor: (query) => {
    const headers = getHeaders();
    return axios.post(burl + "/dispositifs/modifyDispositifMainSponsor", query, {
      headers,
    })
  },

  updateDispositifAdminComments: (query) => {
    const headers = getHeaders();
    return axios.post(burl + "/dispositifs/updateDispositifAdminComments", query, {
      headers,
    })
  },

  createStructure: (query) => {
    const headers = getHeaders();
    return axios.post(burl + "/structures/createStructure", query, {
      headers,
    })
  },

  updateStructure: (query) => {
    const headers = getHeaders();
    return axios.post(burl + "/structures/updateStructure", query, {
      headers,
    })
  },
  modifyUserRoleInStructure: (query) => {
    const headers = getHeaders();
    return axios.post(burl + "/structures/modifyUserRoleInStructure", query, {
      headers,
    })
  },

  getStructureById: (
    id,
    withDisposAssocies,
    localeOfLocalizedDispositifsAssocies,
    withMembres
  ) => {
    const headers = getHeaders();
    return axios.get(burl + "/structures/getStructureById", {
      params: {
        id,
        withDisposAssocies,
        localeOfLocalizedDispositifsAssocies,
        withMembres,
      },
      headers,
    })
  },

  getUserFavoritesInLocale: (locale) => {
    const headers = getHeaders();
    return axios.get(burl + `/user/getUserFavoritesInLocale?locale=${locale}`, {
      headers,
    })
  },
  getNeeds: () =>
    axios.get(burl + "/needs/getNeeds"),

  getDispositifsWithTranslationAvancement: (locale) => {
    const headers = getHeaders();
    return axios.get(
      burl + `/user/getDispositifsWithTranslationAvancement?locale=${locale}`,
      { headers }
    )
  },

  getUserContributions: () => {
    const headers = getHeaders();
    return axios.get(burl + "/user/getUserContributions", { headers })
  },

  updateUserFavorites: (query) => {
    const headers = getHeaders();
    return axios.post(burl + "/user/updateUserFavorites", query, { headers })
  },

  getNbDispositifsByRegion: () =>
    axios.get(burl + "/dispositifs/getNbDispositifsByRegion"),

  getFiguresOnUsers: () =>
    axios.get(burl + "/user/getFiguresOnUsers"),

  exportUsers: () => {
    const headers = getHeaders();
    return axios.post(burl + "/user/exportUsers", {}, { headers })
  },
  exportFiches: () => {
    const headers = getHeaders();
    return axios.post(burl + "/user/exportFiches", {}, { headers })
  },

  exportDispositifsGeolocalisation: () => {
    const headers = getHeaders();
    return axios.post(
      burl + "/dispositifs/exportDispositifsGeolocalisation",
      {},
      { headers }
    )
  },
  getActiveStructures: () =>
    axios.get(burl + "/structures/getActiveStructures"),

  getDispositifs: (params) => {
    const headers = getHeaders();
    return axios.post(burl + "/dispositifs/getDispositifs", params);
  },

  updateNbVuesOrFavoritesOnContent: (params) => {
    const headers = getHeaders();
    return axios.post(burl + "/dispositifs/updateNbVuesOrFavoritesOnContent", params, {
      headers,
    })
  },

  getAllDispositifs: () => axios.get(burl + "/dispositifs/getAllDispositifs"),
  getAllStructures: () => axios.get(burl + "/structures/getAllStructures"),
  getAllUsers: () => {
    const headers = getHeaders();
    return axios.get(burl + "/user/getAllUsers", { headers })
  },

  add_tradForReview: (query) => {
    const headers = getHeaders();
    return axios.post(burl + "/traduction/add_tradForReview", query, {
      headers,
    });
  },
  get_tradForReview: (query) => {
    const headers = getHeaders();
    return axios.post(burl + "/traduction/get_tradForReview", query, {
      headers,
    });
  },
  validateTranslations: (query) => {
    const headers = getHeaders();
    return axios.post(burl + "/traduction/validateTranslations", query, {
      headers,
    });
  },

  saveNeed: (query) =>{
    const headers = getHeaders();
    return axios.post(burl + "/needs/saveNeed", query, {
      headers,
    })
  },

  createNeed: (query) =>{
    const headers = getHeaders();
    return axios.post(burl + "/needs/createNeed", query, {
      headers,
    })
  },

  delete_trads: (query) => {
    const headers = getHeaders();
    return axios.post(burl + "/traduction/delete_trads", query, {
      headers,
    });
  },
  get_progression: (query) => {
    const headers = getHeaders();
    return axios.post(burl + "/traduction/get_progression", query, {
      headers,
    });
  },

  update_tradForReview: (query) => {
    const headers = getHeaders();
    return axios.post(burl + "/traduction/update_tradForReview", query, {
      headers,
    });
  },

  get_translation: (query = {}) => {
    const headers = getHeaders();
    return axios.post(burl + "/translate/get_translation", query, {
      headers,
    });
  },

  get_langues: (query, sort, populate) => {
    const headers = getHeaders();
    return axios.post(
      burl + "/langues/get_langues",
      { query: query, sort: sort, populate: populate },
      { headers }
    );
  },

  getLanguages: () => axios.get(burl + "/langues/getLanguages"),

  set_image: (query) => {
    const headers = getHeaders();
    return axios.post(burl + "/images/set_image", query, { headers });
  },

  set_mail: (query) => {
    const headers = getHeaders();
    return axios.post(burl + "/miscellaneous/set_mail", query, { headers });
  },
  send_sms: (query) => {
    const headers = getHeaders();
    return axios.post(burl + "/miscellaneous/send_sms", query, { headers });
  },

  get_tts: (query) => {
    const headers = getHeaders();
    return axios.post(burl + "/tts/get_tts", query, {
      headers,
      cancelToken: new CancelToken(function executor(c) {
        cancel = c;
      }),
    });
  },
  cancel_tts_subscription: () => cancel && cancel(),

  isAuth: () => {
    if (!isInBrowser()) return false;
    return localStorage.getItem("token") !== null;
  },
  logout: () => {
    setAuthToken(false);
    delete headers["x-access-token"];
    return localStorage.removeItem("token");
  },
};
