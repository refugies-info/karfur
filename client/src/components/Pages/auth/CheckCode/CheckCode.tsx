import React, { useCallback, useMemo, useState } from "react";
import { Col, Row } from "reactstrap";
import { useRouter } from "next/router";
import Link from "next/link";
import Image from "next/image";
import { Button } from "@codegouvfr/react-dsfr/Button";
import { Input } from "@codegouvfr/react-dsfr/Input";
import { useLogin } from "hooks";
import { logger } from "logger";
import API from "utils/API";
import { cls } from "lib/classname";
import isInBrowser from "lib/isInBrowser";
import SEO from "components/Seo";
import GmailIcon from "assets/auth/providers/gmail-icon.svg";
import OutlookIcon from "assets/auth/providers/outlook-icon.svg";
import styles from "scss/components/auth.module.scss";

interface Props {
  type: "2fa" | "login";
}

const CheckCode = ({ type }: Props) => {
  const router = useRouter();
  const email = useMemo(() => router.query.email as string, [router.query]);
  const [error, setError] = useState("");
  const { logUser } = useLogin();

  const submit = useCallback(
    async (e: any) => {
      e.preventDefault();
      const code = e.target.code.value;
      try {
        const res = await API.checkCode({ email, code });
        if (!res.token) throw new Error();
        logUser(res.token);
      } catch (e: any) {
        const errorCode = e.response?.data?.code;
        if (errorCode === "WRONG_CODE") {
          setError("Code incorrect, veuillez réessayer.");
        } else {
          logger.error(e);
          setError("Une erreur s'est produite, veuillez réessayer ou contacter un administrateur.");
        }
      }
    },
    [logUser, email],
  );

  const resendCode = useCallback(
    (e: any) => {
      e.preventDefault();
      API.sendCode({ email });
    },
    [email],
  );

  const contact = useCallback(() => {
    if (!isInBrowser()) return;
    window.$crisp.push(["do", "chat:open"]);
  }, []);

  if (!email) return null;

  return (
    <div className={cls(styles.container, styles.half)}>
      <SEO title="Code de sécurité" />
      <Button priority="tertiary" size="small" iconId="fr-icon-arrow-left-line" onClick={() => router.back()}>
        Retour
      </Button>
      <div className={styles.content}>
        <div className={styles.title}>
          {type === "2fa" ? <h1>Vérifions que c’est bien vous&nbsp;!</h1> : <h1>Entrez le code reçu</h1>}
          <p className={styles.subtitle}>
            Un code temporaire à 6 chiffres vous a été envoyé à {email}
            <br />
            <Link href="/auth" className="text-decoration-underline">
              Ce n'est pas vous&nbsp;?
            </Link>
          </p>
        </div>

        <form onSubmit={submit}>
          <Input
            label={type === "2fa" ? "Code de vérification" : "Code de connexion temporaire"}
            state={!error ? "default" : "error"}
            stateRelatedMessage={error}
            nativeInputProps={{
              autoFocus: true,
              name: "code",
            }}
          />

          <Button
            iconId="fr-icon-check-line"
            iconPosition="right"
            className={cls(styles.button, "mt-8 mb-4")}
            nativeButtonProps={{ type: "submit" }}
          >
            Valider
          </Button>
          <Button
            iconId="fr-icon-mail-line"
            iconPosition="right"
            onClick={resendCode}
            className={styles.button}
            priority="tertiary"
          >
            Renvoyer le code
          </Button>
        </form>

        {type === "2fa" && (
          <div className={cls(styles.small, "mt-14 mb-14", "text-center")}>
            L'adresse mail n'est plus valable&nbsp;?{" "}
            <button onClick={contact} className={styles.link}>
              Contactez-nous
            </button>
          </div>
        )}

        <Row className={cls("mb-4", type !== "2fa" && styles.space_top)}>
          <Col>
            <Button
              linkProps={{
                href: "https://mail.google.com/mail/u/0/",
              }}
              className={styles.button}
              priority="tertiary"
            >
              <Image src={GmailIcon} width={24} height={24} alt="" className="me-2" />
              Ouvrir Gmail
            </Button>
          </Col>
          <Col>
            <Button
              linkProps={{
                href: "https://outlook.live.com/mail/0/",
              }}
              className={styles.button}
              priority="tertiary"
            >
              <Image src={OutlookIcon} width={24} height={24} alt="" className="me-2" />
              Ouvrir Outlook
            </Button>
          </Col>
        </Row>
        <p className={cls(styles.small, "mt-6", "text-center")}>Pensez à vérifiez votre courrier indésirable&nbsp;!</p>
      </div>
    </div>
  );
};

export default CheckCode;
