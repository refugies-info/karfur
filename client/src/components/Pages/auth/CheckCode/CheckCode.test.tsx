import { fireEvent, waitFor } from "@testing-library/react";
import mockRouter from "next-router-mock";
import API from "utils/API";
import { initialMockStore } from "__fixtures__/reduxStore";
import { wrapWithProvidersAndRenderForTesting } from "../../../../../jest/lib/wrapWithProvidersAndRender";
import CheckCode from "./CheckCode";

jest.mock("next/router", () => require("next-router-mock"));
const logUserMock = jest.fn();

jest.mock("hooks/useLogin", () => {
  return () => ({
    logUser: logUserMock,
    handleError: jest.fn(),
  });
});

describe("CheckCode", () => {
  // Renders the component with the correct title and subtitle
  it("should render the component with the correct title and subtitle", async () => {
    await mockRouter.push("/check-code?email=test@example.com");

    const { asFragment } = wrapWithProvidersAndRenderForTesting({
      Component: CheckCode,
      reduxState: {
        ...initialMockStore,
      },
      compProps: {
        type: "2fa",
      },
    });

    expect(asFragment()).toMatchSnapshot();
  });

  // Does not render the component if the email is not provided
  it("should not render the component if the email is not provided", async () => {
    await mockRouter.push("/check-code");
    const { asFragment } = wrapWithProvidersAndRenderForTesting({
      Component: CheckCode,
      reduxState: {
        ...initialMockStore,
      },
      compProps: {
        type: "2fa",
      },
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it("should display an error message if code is wrong", async () => {
    await mockRouter.push("/check-code?email=test@example.com");

    jest.spyOn(API, "checkCode").mockRejectedValueOnce({ response: { data: { code: "WRONG_CODE" } } });

    const component = wrapWithProvidersAndRenderForTesting({
      Component: CheckCode,
      reduxState: {
        ...initialMockStore,
      },
      compProps: {
        type: "2fa",
      },
    });

    fireEvent.change(component.getByRole("textbox"), { target: { value: "12345" } });
    fireEvent.click(component.getByText("Valider"));

    await waitFor(() => expect(component.getByText("Code incorrect, veuillez réessayer.")).toBeTruthy());
  });

  it("should display an error message if an API error occurs", async () => {
    mockRouter.push("/check-code?email=test@example.com");

    jest.spyOn(API, "checkCode").mockRejectedValueOnce(new Error());
    const component = wrapWithProvidersAndRenderForTesting({
      Component: CheckCode,
      reduxState: {
        ...initialMockStore,
      },
      compProps: {
        type: "2fa",
      },
    });

    fireEvent.change(component.getByRole("textbox"), { target: { value: "12345" } });
    fireEvent.click(component.getByText("Valider"));

    await waitFor(() =>
      expect(
        component.getByText("Une erreur s'est produite, veuillez réessayer ou contacter un administrateur."),
      ).toBeTruthy(),
    );
  });

  it("should display an error message if no token is returned", async () => {
    await mockRouter.push("/check-code?email=test@example.com");

    //@ts-expect-error
    jest.spyOn(API, "checkCode").mockResolvedValue({});
    const component = wrapWithProvidersAndRenderForTesting({
      Component: CheckCode,
      reduxState: {
        ...initialMockStore,
      },
      compProps: {
        type: "2fa",
      },
    });

    fireEvent.change(component.getByRole("textbox"), { target: { value: "12345" } });
    fireEvent.click(component.getByText("Valider"));

    await waitFor(() =>
      expect(
        component.getByText("Une erreur s'est produite, veuillez réessayer ou contacter un administrateur."),
      ).toBeTruthy(),
    );
  });

  it("should call logUser if all is ok", async () => {
    await mockRouter.push("/check-code?email=test@example.com");

    jest.spyOn(API, "checkCode").mockResolvedValue({ token: "ok" });
    const component = wrapWithProvidersAndRenderForTesting({
      Component: CheckCode,
      reduxState: {
        ...initialMockStore,
      },
      compProps: {
        type: "2fa",
      },
    });

    fireEvent.change(component.getByRole("textbox"), { target: { value: "12345" } });
    fireEvent.click(component.getByText("Valider"));
    await waitFor(() => expect(logUserMock).toHaveBeenCalled());
  });
});
