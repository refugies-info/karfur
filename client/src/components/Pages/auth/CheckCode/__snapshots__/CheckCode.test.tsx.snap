// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckCode should not render the component if the email is not provided 1`] = `null`;

exports[`CheckCode should render the component with the correct title and subtitle 1`] = `null`;
